### PR TITLE
Better validation

### PR DESCRIPTION
**Pull request by anlthms**
this is mainly to get a more reliable validation loss history. skipping bprop periodically is hacky and was found to break in two cases:
1) if the training data is reshuffled on every epoch, we end up validating on different batches 
2) if there are dependencies between samples (e.g. patches from the same image), could lead to leakage of training data into validation

to fix, an explicit validation set is used that is fixed for the entire training process.

### CODE REVIEW

**Review by rgharris — APPROVED** [review]
Makes sense and looks good 👍. 

Was the original motivation for skipping backprop to make it easier to "early rankify" user code (requiring less change to user code)?

**Comment by anlthms** [comment]
> Was the original motivation for skipping backprop to make it easier to "early rankify" user code (requiring less change to user code)?

the original idea was to collect history with zero changes to user code! a shim layer that sits on top of pytorch would have intercepted the bprop calls and swallowed some of them, while passing the rest of them through to pytorch. but that has the two issues mentioned above.

the current solution is a bit clunky, but reliable.

**Comment by rgharris** [comment]
👍 thanks for the clarification, all makes sense! Makes sense to close these gaps and we can work on making the user setup easier down the road.
